File: webbluetooth/packages/brilliant-msg/example/live-camera-feed.js
import { FrameMsg, StdLua, TxCaptureSettings, RxPhoto } from 'frame-msg';
import { BrilliantDeviceType } from 'frame-ble';
import frameApp from './lua/live_camera_feed_frame_app.lua?raw';

/**
 * Demonstrates capturing a sequence of photos from the Frame camera and displaying them
 * on a webpage to create a live feed effect.
 * This example involves:
 * - Iteratively requesting photos from the Frame device using `TxCaptureSettings`.
 * - Receiving the JPEG image data via `RxPhoto`.
 * - Displaying each newly captured photo in an HTML image element, replacing the previous one.
 */
export async function run() {
  const frame = new FrameMsg();

  try {
    // Web Bluetooth API requires a user gesture to initiate the connection
    // This is usually a button click or similar event
    console.log("Connecting to Frame...");
    const deviceId = await frame.connect();
    console.log('Connected to:', deviceId);

    // debug only: check our current battery level and memory usage (which varies between 16kb and 31kb or so even after the VM init)
    const battMem = await frame.sendLua('print(frame.battery_level() .. " / " .. collectgarbage("count"))', {awaitPrint: true});
    console.log(`Battery Level/Memory used: ${battMem}`);

    // Let the user know we're starting
    await frame.printShortText('Loading...');

    // send the std lua files to Frame
    await frame.uploadStdLuaLibs([StdLua.DataMin, StdLua.CameraMin]);

    // Send the main lua application from this project to Frame that will run the app
    await frame.uploadFrameApp(frameApp);

    // attach the print response handler so we can see stdout from Frame Lua print() statements
    // If we assigned this handler before the frameside app was running,
    // any await_print=True commands will echo the acknowledgement byte (e.g. "1"), but if we assign
    // the handler now we'll see any lua exceptions (or stdout print statements)
    frame.attachPrintResponseHandler(console.log);

    // "require" the main frame_app lua file to run it, and block until it has started.
    // It signals that it is ready by sending something on the string response channel.
    await frame.startFrameApp();

    // hook up the RxPhoto receiver
    const upright = frame.ble.type === BrilliantDeviceType.FRAME; // rotate images from Frame 90 degrees counterclockwise, not Halo
    const rxPhoto = new RxPhoto({upright: upright});
    const photoQueue = await rxPhoto.attach(frame);

    // create the element to display the photo
    const img = document.createElement('img');
    const imageDiv = document.getElementById('image1');
    if (imageDiv) {
      // Clear any existing content in the div
      while (imageDiv.firstChild) {
        imageDiv.removeChild(imageDiv.firstChild);
      }
      imageDiv.appendChild(img);
    }

    // loop 20 times - take a photo and display it in the div
    for (let i = 0; i < 20; i++) {
      // Request the photo by sending a TxCaptureSettings message
      const resolution = frame.ble.type === BrilliantDeviceType.FRAME ? 512 : 640;
      await frame.sendMessage(0x0d, new TxCaptureSettings({resolution: resolution}).pack());

      // get the jpeg bytes as soon as they're ready
      const jpegBytes = await photoQueue.get();
      console.log("Photo received, length:", jpegBytes.length);

      // display the image on the web page
      // overwriting the previous image
      img.src = URL.createObjectURL(new Blob([jpegBytes], { type: 'image/jpeg' }));
    }

    // stop the photo listener and clean up its resources
    rxPhoto.detach(frame);

    // unhook the print handler
    frame.detachPrintResponseHandler()

    // break out of the frame app loop and reboot Frame
    await frame.stopFrameApp()
  }
  catch (error) {
    console.error("Error:", error);
  }
  finally {
    // Ensure the Frame is disconnected in case of an error
    try {
      await frame.disconnect();
      console.log("Disconnected from Frame.");
    } catch (disconnectError) {
      console.error("Error during disconnection:", disconnectError);
    }
  }
};

File: python/packages/brilliant_msg/examples/live_camera_feed.py
import asyncio
import argparse
from PIL import Image
import io
import cv2
import numpy as np

from brilliant_msg import BrilliantMsg, RxPhoto, TxCaptureSettings
from brilliant_ble import BrilliantDeviceType

async def main():
    """
    Take photos continuously using the Frame camera and display them in an OpenCV window
    """
    parser = argparse.ArgumentParser(description="Connect to a Halo/Frame device and run this example.")
    parser.add_argument(
        "--name",
        default=None,
        help='exact BLE device name, e.g. "Halo AB" or "Frame 4F"; defaults to the nearest device',
    )
    args = parser.parse_args()
    frame = None
    rx_photo = None
    window_name = "Camera Feed"
    
    try:
        # Initialize OpenCV Window on Main Thread
        cv2.namedWindow(window_name, cv2.WINDOW_NORMAL)
        
        frame = BrilliantMsg()
        name = await frame.connect(name=args.name)
        fw = await frame.send_lua("print(frame.FIRMWARE_VERSION)", await_print=True)
        tag = await frame.send_lua("print(frame.GIT_TAG)", await_print=True)
        batt = await frame.send_lua("print(frame.battery_level())", await_print=True)
        print(f"{name} | firmware {fw} | git {tag} | battery {batt}%")

        # debug only: check our current battery level and memory usage
        batt_mem = await frame.send_lua('print(frame.battery_level() .. " / " .. collectgarbage("count"))', await_print=True)
        print(f"Battery Level/Memory used: {batt_mem}")

        # Let the user know we're starting
        await frame.print_short_text('Loading...')

        # send the std lua files to Frame
        await frame.upload_stdlua_libs(lib_names=['data', 'camera'])

        # Send the main lua application
        await frame.upload_frame_app(local_filename="lua/camera_frame_app.lua")

        frame.attach_print_response_handler()

        # Start the frame app
        await frame.start_frame_app()

        # hook up the RxPhoto receiver
        rx_photo = RxPhoto(upright=frame.ble.type == BrilliantDeviceType.FRAME)
        photo_queue = await rx_photo.attach(frame)

        if frame.ble.type == BrilliantDeviceType.FRAME:
            # give the frame some time for the autoexposure loop to run
            print("Letting autoexposure loop run for 5 seconds to settle")
            
            # We process the UI loop during the sleep to prevent "spinning wheel"
            for _ in range(50):
                await asyncio.sleep(0.1)
                cv2.waitKey(1)

        print("Starting continuous capture - Ctrl+C to stop")

        # Main capture loop
        capture_count = 0
        while True:
            # Check if window was closed via 'X' button
            if cv2.getWindowProperty(window_name, cv2.WND_PROP_VISIBLE) < 1:
                print("Window closed, exiting...")
                break

            # Request a photo
            resolution: int = 720 if frame.ble.type == BrilliantDeviceType.FRAME else 640
            await frame.send_message(0x0d, TxCaptureSettings(resolution=resolution).pack())
            
            try:
                # Wait for image with short timeout to keep UI responsive
                # If the image takes > 0.1s, we catch TimeoutError, update UI, and wait again
                jpeg_bytes = await asyncio.wait_for(photo_queue.get(), timeout=0.1)
                
                # Convert PIL Image to OpenCV format
                pil_image = Image.open(io.BytesIO(jpeg_bytes))
                cv_image = np.array(pil_image)
                
                # Convert RGB to BGR (OpenCV uses BGR)
                if cv_image.shape[2] == 3:
                    cv_image = cv2.cvtColor(cv_image, cv2.COLOR_RGB2BGR)
                
                # Display image
                cv2.imshow(window_name, cv_image)
                
                capture_count += 1
                print(f"Captured frame {capture_count}", end="\r")

            except asyncio.TimeoutError:
                # No frame arrived yet, just continue to keep UI alive
                pass
            
            # Pump OpenCV events (Required for Mac GUI)
            # waitKey(1) blocks for 1ms. 
            key = cv2.waitKey(1) & 0xFF
            if key == 27:  # ESC key
                break
            
            # Small yield to let other async tasks run
            await asyncio.sleep(0.01)
            
    except asyncio.CancelledError:
        print("\nCapture loop cancelled")
    except Exception as e:
        print(f"\nAn error occurred: {e}")
    finally:
        # Clean up resources
        print("\nCleaning up resources...")
        cv2.destroyAllWindows()
        if rx_photo and frame:
            rx_photo.detach(frame)
        if frame:
            frame.detach_print_response_handler()
            await frame.stop_frame_app()
            await frame.disconnect()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nProgram interrupted by user")